### DISCUSSION TITLE
Importer not working - networking/environment variable issue?

### DISCUSSION THREAD

**Discussion by AlteVerwischteZiege** [firefly-iii/firefly-iii#9537] [Firefly III]
I am having substantially the same issue, with the reported solution already implemented.

Do begin with, these are my differences with  **Frittenpause**'s  instalation:

- I am running a Docker container on a Synology NAS.
- I followed the directions from this YouTube video: https://www.youtube.com/watch?v=quJgB0STYac
- The IP of my VM is 192.168.9.52:901 and the Importer at 192.168.9.52:902
- The Importer callback is http://192.168.9.52:902/callback

I too can successfully log into Firefly III at http://192.168.9.52:901. and the importer at http://192.168.9.52:901.

The importer is generating the called URL after providing the client ID of:  http://[localhost](http://localhost/oauth/authorize?client_id=4&redirect_uri=http%3A%2F%2F192.168.9.52%3A902%2Fcallback&response_type=code&scope=&state=BzBjxWR1TORBt2X3ZrfxDfWNaRz9zQBqexN7pHtp&code_challenge=kd1WG2IIhdEPL4QzHQUAtIlqEuvkAijOg4iNSY44Ubw&code_challenge_method=S256)/oauth/authorize?client_id=4&redirect_uri=http%3A%2F%2F192.168.9.52%3A902%2Fcallback&response_type=code&scope=&state=BzBjxWR1TORBt2X3ZrfxDfWNaRz9zQBqexN7pHtp&code_challenge=kd1WG2IIhdEPL4QzHQUAtIlqEuvkAijOg4iNSY44Ubw&code_challenge_method=S256

This produces an error message of "localhost refused to connect."

If I manually update the called URL, replacing the "localhost" with 192.168.9.52:901, like this:
192.168.9.52:901/oauth/authorize?client_id=4&redirect_uri=http%3A%2F%2F192.168.9.52%3A902%2Fcallback&response_type=code&scope=&state=BzBjxWR1TORBt2X3ZrfxDfWNaRz9zQBqexN7pHtp&code_challenge=kd1WG2IIhdEPL4QzHQUAtIlqEuvkAijOg4iNSY44Ubw&code_challenge_method=S256

The importer launches successfully.

My .importer.env contains the following configuration entries:
FIREFLY_III_URL=http://192.168.9.52:901
VANITY_URL=http://192.168.9.52:901

Thoughts?

### Discussed in https://github.com/orgs/firefly-iii/discussions/8970

<div type='discussions-op-text'>

<sup>Originally posted by **Frittenpause** June 14, 2024</sup>
Hello everyone,

I decided to ask here after my normal resources and searches have not yet gotten me anywhere. I am currently in the process of setting up my Server at home and, after some research, decided to go for Firefly because it seemed to be the nicest system.

I am running a Proxmox and one VM has Ubuntu server on it with Portainer. I followed the instructions to get it up and running. (https://docs.firefly-iii.org/how-to/firefly-iii/installation/portainer/)

I can log into the firefly iii system and the importer is successfully installed and giving me the startup screen.
<img width="937" alt="Screenshot 2024-06-14 at 15 14 35" src="https://github.com/firefly-iii/firefly-iii/assets/4592181/9465c8f5-f9fd-4fda-99b7-c2ef26253dc6">

The IP of the VM is 192.168.10.109, the port for Firefly is 81, for the importer 82. The Container IPs are as seen in the screenshot. 
I did try to follow the steps and created a client ID with the  callback http://192.168.10.109:82/callback but to no avail. I also tried the Container IPs and even localhost (which is said to be not working on docker according to the documentation).

The called URL when i try to enter the client ID in the importer is 

http://localhost/oauth/authorize?client_id=14&redirect_uri=http%3A%2F%2F192.168.10.109%3A82%2Fcallback&response_type=code&scope=&state=[...]&code_challenge=[...]&code_challenge_method=S256

this fails wit "Unable to connect".

In the Importer Container Logs i can find this:

> 2024-06-14 15:31:25] production.DEBUG: Now at App\Http\Controllers\TokenController::submitClientId  
> 
> [2024-06-14 15:31:25] production.DEBUG: Submitted data:  {"client_id":"14"} 
> 
> [2024-06-14 15:31:25] production.DEBUG: [a] Base URL is "http://app:8080"  
> 
> [2024-06-14 15:31:25] production.DEBUG: [b] Vanity URL is now "http://app:8080"  
> 
> [2024-06-14 15:31:25] production.DEBUG: [c] Vanity URL is now "http://localhost"  
> 
> [2024-06-14 15:31:25] production.DEBUG: Now in App\Http\Controllers\TokenController::redirectForPermission(request, "http://app:8080", "http://localhost", 14)  
> 
> [2024-06-14 15:31:25] production.DEBUG: Query parameters are {"client_id":14,"redirect_uri":"http://192.168.10.109:82/callback","response_type":"code","scope":"","state":"[...]","code_challenge":"-[...]","code_challenge_method":"S256"}

There are no logs on the  firefly iii container at the same time.

As far as I understand the importer tries to direct me to localhost and from there to the external IP outside the container. The IP redirection does not change, no matter what I entered as the redirect URL when creating the client in the firefly instance.

I am think I might have missed something while creating the Stack and have now some Variables pointing at the wrong address, but I am not able to pinpoint it. 

The system is completely new setup and I can easily redo the whole process, I just would like to know what I did wrong before I do so.

Thanks a lot and greetings from Japan

Nils
</div>

**Comment by JC5** [comment]
Mmmm, I'm surprised it sends you to localhost, especially since you set the IP yourself.

Try browsing to `/flush`, it may reset the cookies that also hold these variables.

**Reply by AlteVerwischteZiege** [reply]
> Mmmm, I'm surprised it sends you to localhost, especially since you set the IP yourself.
> 
> Try browsing to `/flush`, it may reset the cookies that also hold these variables.

I am not familiar with `/flush` (bear with me, I am learning).  If your instruction is to enter `/flush` as the URL, I have done so with no change to the result.

**Reply by JC5** [reply]
Mmm, that was indeed what I meant, but other than that, it's hard to be sure what it could be.

**Reply by FUF1** [reply]
Maybe it helps in finding the issue: I've got the ERROR CONNECTION_REFUSED when setting up firefly data importer as per the guide with the localhost vanity URL and the OAuth token. Thanks to your Q&A and another post I hardcoded the VANITY_URL in my environment variables and added the port of the firefly installation (not really knowing what I am doing to be honest). For now it seems to work. (APP_URL and FIREFLY_III_URL are still the same as in the setup guide.)
